### PR TITLE
feat: add `indexPrefix` option

### LINKED ISSUE #162: Regression in v4.0.0 when hosted behind a path rewriting reverse proxy
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the regression has not already been reported

### Last working version

3.1.0

### Stopped working in version

4.0.0

### Node.js version

18.x

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

Debian 11

### 💥 Regression Report

When a reverse proxy rewrites the URL path before forwarding the request to the Node.js server running swagger-ui the interface gets confused and the Swagger UI becomes inoperable.

For example when a `GET /api/documentation` request is remapped to `GET /documentation` before being passed on to swagger-ui v4.0.0 the resulting html page it responds to contains resources links to `/documentation/static/<resource>`, which will not load as the server hosts these resources under `/api/documentation/static/<resource>`. Trying to access `/api/documentation/static/index.html` which used to work causes swagger-ui to redirect the browser to `/documentation`.

In swagger-ui v3.1.0 all links were relative. A `GET /api/documentation` request remapped to `GET /documentation` would correctly reply with a relative redirect to `./documentation/static/index.html` sending the browser to `/api/documentation/static/index.html`. This page only used relative links of the form `./<resource>` so it wasn't sensitive to the path prefix being remapped.

### Steps to Reproduce

Configure a reverse proxy in front of swagger-ui, for example with an Apache ProxyPass directive.

```apache
<Location /api/>
    ProxyPass http://localhost:3000/
</Location>
```

Observe that in v3.1.0 you can access the swagger UI under /api/documentation, while in v4.0.0 it doesn't work at all.

### Expected Behavior

Upgrading from 3.1.0 shouldn't cause the swagger UI to break when hosted behind a reverse proxy where the path has been rewritten. Or alternatively, this breakage/new requirement should be clearly documented.

### LINKED ISSUE #176: Can't specify host to work correctly with nginx generated subroutes
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the regression has not already been reported

### Last working version

7.4.0

### Stopped working in version

9.0.0

### Node.js version

22.9.0

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

14.0

### 💥 Regression Report

I'm using fastify with docker and nginx API gateway, so I always need to specify service api url in my swagger configuration, so that it is routed correctly. 
after update, I couldn't find "host" parameter I used before, and while it all works fine locally, it crashes on server, being unable to fetch lib files, such as swagger-ui-bundle.js.
e.g. request goes to https://mydomain/docs/swagger-ui-bundle.js while the correct path is https://mydomain/v1/mock-server/docs/swagger-ui-bundle.js and getting either html or 404 response

routePrefix and servers parameters couldn't help me here 

maybe I'm missing an obvious solution, but I couldn't find any in documentation or existing tickets.

sources mismatch:
photo_2024-09-24 13 49 58

thanks in advance!

### Steps to Reproduce

clone repo 
https://github.com/Danko14/fastify-swagger

follow readme steps (create .env & run docker-compose)
you'll get nginx running on 80 and mock server running on 8080 and available on http://localhost/mock-server/
query http://localhost/mock-server/ping to get { msg: 'ok' }
query  http://localhost:8080/docs to see swagger docs
query  http://localhost/mock-server/docs to see where the problem is

swagger config:
` fastify.register(
        fastifySwagger,
        {
            openapi: {
                info: {
                    title: 'Mock server',
                    description: 'API documentation',
                    version,
                },
                servers: [
                    { url: `https://${process.env.SERVER_API_URL}`, description: 'HTTPS' },
                    { url: `http://${process.env.HTTP_HOST}:${process.env.HTTP_PORT}`, description: 'HTTP' },
                ],
                components: {
                    securitySchemes: {
                        actorId: {
                            type: 'apiKey',
                            in: 'header',
                            name: 'x-actor-id',
                        },
                        actorType: {
                            type: 'apiKey',
                            in: 'header',
                            name: 'x-actor-type',
                        },
                    },
                },
                security: [
                    { ApiKeyAuth: [] },
                    { ActorTypeAuth: [] },
                ],
            },
        },
    )
    fastify.register(swaggerUi, {
        routePrefix: '/docs',
        uiConfig: {
            docExpansion: 'list',
            deepLinking: false,
        },
        staticCSP: false,
        transformSpecification: (swaggerObject, request, reply) => swaggerObject,
        transformSpecificationClone: true,
    })`
    
nginx config: 
image

### Expected Behavior

_No response_

### LINKED ISSUE #180: Issue with index html resources paths when using a complex path as route prefix
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

5.0.0

### Plugin version

5.1.0

### Node.js version

20.18

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

15.0.1

### Description

It's appears [#178](https://github.com/fastify/fastify-swagger-ui/pull/178) is causing an issue when using a complex path in `routePrefix`, ie:

```ts
await this.instance.register(fastifySwaggerUI, {
  routePrefix: '/api/vi/$docs'
});
```

The resolved path looks like `/api/v1/api/v1/$docs/static` in that case.

image

See linked repo for MRE.

### Link to code that reproduces the bug

https://github.com/jpb06/fastify-swagger-ui-routeprefix-issue-repro

### Expected Behavior

index resources to load properly.

### PR DESCRIPTION
**Pull request by DavidTanner**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Adds a feature flag for #176, #164, #162, and #180

### CODE REVIEW

**Comment by DavidTanner** [comment]
@mcollina  could you please review this?

File: types/index.d.ts [hunk 1]
```diff
@@ -45,6 +45,10 @@ declare namespace fastifySwaggerUi {
      * @default /documentation
      */
     routePrefix?: string;
+    /**
+     * Add an index prefix. This is for when the Fastify server is behind path based routing.  ex. NGINX
+     */
+    indexPrefix?: string;
```
Comment (mcollina) [inline types/index.d.ts:51]: Can you add a test for this type? We use tsd.

**Review by mcollina — COMMENTED** [review]
Good work!

File: types/index.d.ts (hunk 1 above)
Comment (DavidTanner) [reply types/index.d.ts:51]: I'm not sure what you mean.  There is a test in [test/route.test.js,](https://github.com/fastify/fastify-swagger-ui/pull/186/files#diff-71868d2908a96fcd854d3015e6275d3841daa70478fa84f46661f23da563c802R585-R616)   Is there something else you are looking for?

**Comment by balanced-hvu** [comment]
Hi everyone,

Have we reached a consensus and a release date? I'm also waiting and want to remove a similar hack (inspired by this PR) that we've implemented to address the relative asset paths problem.

Many thanks!

File: types/index.d.ts (hunk 1 above)
Comment (rozzilla) [reply types/index.d.ts:51]: Hi @DavidTanner! 👋🏼 
You can see an example [here](https://github.com/fastify/fastify-swagger-ui/blob/master/types/types.test-d.ts). Basically, by running the `tsd` tool, you can run tests for the TS types. You can then check locally by running `npm run typescript`, (which [will run](https://github.com/fastify/fastify-swagger-ui/blob/master/package.json#L18) `tsd`)

**Review by rozzilla — APPROVED** [review]

File: types/index.d.ts (hunk 1 above)
Comment (rozzilla) [reply types/index.d.ts:51]: Happy to push the TS test update if this can help to unlock the release of this PR 😉

Comment (rozzilla) [reply types/index.d.ts:51]: https://github.com/fastify/fastify-swagger-ui/pull/189/commits

**Pull request closed without merging**